Decouple flux communication from Metavariables

nilsvu opened the pull request:
## Proposed changes

This PR decouples the actions that send, receive and apply fluxes from `Metavariables`, and instead templates them only on the information they need (generally this is the `TemporalIdTag` and the type that computes the flux). Aside from making the code less dependent on global information in `Metavariables` or `Metavariables::system` (which I think we should generally strive for, see #797), this allows us to use the infrastructure to communicate multiple fluxes and/or fluxes for a subset of variables, e.g. to compute an auxiliary field and the corresponding physical field in two separate steps with different fluxes.

### Types of changes:

- [x] New feature

### Component:

- [x] Code

### Code review checklist

- [ ] The PR passes all checks, including unit tests, `clang-tidy` and `IWYU`. For instructions on how to perform the CI checks locally refer to the [Dev guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`. Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the [code review guide](https://spectre-code.org/code_review_guide.html).

nilsdeppe reviewed (changes requested):
Before this gets reviewed please minimize the changes. For example, you are adding `evolve_algorithm_actions`. Is this necessary here? Why is it being done? Does this have anything at all to do with flux communication? If not, don't change it because I just spent time trying to figure out how it's used.

nilsdeppe commented on src/Evolution/DiscontinuousGalerkin/InitializeElement.hpp:
```diff
@@ -90,6 +90,16 @@ struct InitializeElement {
   template <typename Tag>
   using boundary_tag = Tags::Interface<Tags::BoundaryDirections<Dim>, Tag>;
 
+  // It would be good to decouple this from the Metavariables entirely. This
```
Why would this be good? You say what you think is necessary to do it but I have no idea why you want to do it.

nilsdeppe commented on src/Evolution/Executables/ScalarWave/EvolveScalarWave.hpp:
```diff
@@ -35,37 +35,40 @@
 template <size_t Dim>
 struct EvolutionMetavars {
+ private:
+  using temporal_id_tag = Tags::TimeId;
```
why is this private? It'll be necessary for slope limiters too.

nilsdeppe commented on src/Evolution/Executables/ScalarWave/EvolveScalarWave.hpp:
```diff
@@ -35,37 +35,40 @@
   using analytic_solution_tag =
       CacheTags::AnalyticSolution<ScalarWave::Solutions::PlaneWave<Dim>>;
-  using normal_dot_numerical_flux =
+  using numerical_flux_tag =
```
Why are you changing the name of this? Is it important? The old name actually describes what it is, the new one is incorrect.

nilsdeppe commented on src/Evolution/Executables/ScalarWave/EvolveScalarWave.hpp:
```diff
@@ -35,37 +35,40 @@
-    Evolve,
-    Exit
-  };
+  enum class Phase { Initialization, Evolve, Exit };
```
again, try to avoid these types of changes in commits that require us to understand the code in detail.

nilsdeppe commented on src/Evolution/Systems/ScalarWave/Equations.hpp:
```diff
@@ -86,6 +86,7 @@ struct ComputeDuDt {
 template <size_t Dim>
 struct ComputeNormalDotFluxes {
   using argument_tags = tmpl::list<Pi, Phi<Dim>>;
+  using target_fields = tmpl::list<Pi, Phi<Dim>, Psi>;
```
Why is this better than supplying the `package_tags`?

wthrowe replied:
I don't understand this question.  There is nothing called `package_tags` here and there is no operation going on that I would consider packaging.

nilsdeppe replied:
```
  using package_tags =
      tmpl::list<Tags::NormalDotFlux<Pi>, Tags::NormalDotFlux<Phi<Dim>>, Pi,
                 NormalTimesFluxPi>;
```
line 133 of the Equations.hpp in develop

Edit: Github hid the lines so nicely... That's in a different class...

I still don't understand how `target_fields` has anything to do with the arguments passed here. It's not the tags that are passed and so no guarantee the types are correct... It seems really strange to me.

nilsvu replied:
I agree the changes in this PR are not so obvious, it's probably better to talk about it in a hangout. We discussed this in a hangout yesterday (@wthrowe, @kidder, @osheamonn and I) and I'm happy to talk to you too when you have time. I'm also going to add some documentation to provide a better basis for discussion than the raw code.

For now a short walkthrough:

- `target_fields` specifies the databox tags that a numerical flux computer applies to. That means its `()` operator is expected to compute `db::wrap_tags_in<Tags::NormalDotNumericalFlux, target_fields>`.
- To do that, the `()` operator needs the `package_tags` on both sides of the interface.
- The packaging, in turn, needs the `slice_tags` (to be renamed to `argument_tags`) on the interface, plus the unit normal.

The flux communication makes the packaged data available on both sides of the interface. Then in the action where the fluxes are applied, the `()` operator is called. Usually, the computed numerical fluxes F^star are then passed into `lift_flux`, and this is where also the N.F are needed since it computes the strong form -N.(F^star - F). We know which N.F we need here, namely `db::wrap_tags_in<Tags::NormalDotFlux, target_fields>`, so these are extracted from the databox.

nilsdeppe commented on src/Evolution/Systems/ScalarWave/Equations.hpp:
```diff
@@ -130,25 +131,41 @@ struct UpwindFlux {
+  // These tags are sliced to the interface of the element and passed to
+  // `package_data` to provide the data needed to compute the numerical fluxes.
+  using slice_tags =
+      tmpl::list<Tags::NormalDotFlux<Pi>, Tags::NormalDotFlux<Phi<Dim>>, Pi>;
```
N.F is NOT sliced to the boundary because we don't have it in the volume. This makes absolutely no sense for hyperbolic systems.

wthrowe replied:
We should rename the list anyway (argument_tags like most other places?) since it isn't important that they are obtained by slicing.  In recent discussion about boundary conditions it has come up that we may want to use these functions, and it that case values will be directly computed on the boundary and not sliced.

nilsdeppe reviewed (commented):
@kidder asked me to list some questions I have on this so here they are. The rest of this seems propagating the changes

nilsvu committed:
Decouple flux communication from Metavariables and System

nilsvu commented:
@nilsdeppe As you suggested I minimized the changes, removed unnecessary reformatting. I also fixed the tests, so you can see the effect this PR has on the architecture e.g. in `Test_FluxCommunication.hpp`: In that test we don't even need to define a `System` struct anymore, and of the `Metavariables` remains only what the `ActionRunner` needs. This means the flux communication is independent of the physical system and only needs to know the fluxes to communicate.

If the tests and the documentation in `FluxCommunicationTypes.hpp` don't answer your questions sufficiently let's talk in a hangout, since that's probably going to be more efficient.

nilsvu commented:
@wthrowe This update pretty much finalizes what I had in mind for the flux communication, could you look at this and give me your feedback?

wthrowe commented on src/NumericalAlgorithms/DiscontinuousGalerkin/Actions/FluxCommunication.hpp:
```diff
@@ -174,34 +170,33 @@ struct ReceiveDataForFluxes {
+///   - Interface<Tags::Normalized<Tags::UnnormalizedFaceNormal<Dim>>>
+///   - Interface<Tags::Magnitude<Tags::UnnormalizedFaceNormal<Dim>>>,
+///   - TemporalIdTag
 ///   - Tags::Next<Metavariables::temporal_id>
```
Missed one

wthrowe commented on src/Evolution/Systems/ScalarWave/Equations.hpp:
```diff
@@ -130,25 +131,41 @@ struct UpwindFlux {
+  // These tags on the interface of the element are passed to
+  // `package_data` to provide the data needed to compute the numerical fluxes.
+  using argument_tags =
+      tmpl::list<Tags::NormalDotFlux<Pi>, Tags::NormalDotFlux<Phi<Dim>>, Pi>;
```
Since we're getting rid of the implicit n.f arguments, let's get rid of the implicit face normal argument as well and put everything into the argument tags.

kidder replied:
@nilsleiffischer can you make a small PR just adding the comments in this file and the change of `slice_tags`

nilsvu replied:
done in #820

wthrowe reviewed (changes requested):
The changes to the flux communication look like a reasonable way to decouple things.  However, as we discussed in our meeting, it is not clear if this is necessary.  So we still need to determine that we need it or that it is desirable even if not necessary before we can really move ahead here.

I do think we should merge the changes to `slice_tags`.  You can make a separate PR for that if you want to get it in more quickly, or we can just wait to see what we decide here.

nilsvu commented:
Ok so these are the issues to discuss:
- [x] Rename `slice_tags` to `argument_tags` and getting rid of the implicit arguments passed to `package_tags`.
- [ ] Remove the dependence of the flux communication on `System`. I think that's quite desirable independent of whether it's necessary for the elliptic solver. In particular it makes the action list for the algorithm in the Metavariables much more expressive since there are no (less) hidden dependency injections. What's your take on that @wthrowe?

wthrowe commented:
There has been some talk of allowing access to the ConstGlobalCache through the DataBox, and if we do that and stop passing it to the actions they will lose their implicit knowledge of the metavariables.  Then we will have to template most actions on something (and it may be desirable independent of that).  I think there are three reasonable paths in that case:
1. Template on the metavariables, effectively the same as now
2. Template on metavariables entries and the System, as needed
3. Template on the exact parameters needed, whether from the metavariables directly or from the System.

This PR tends towards 3, I think I prefer 2.  Anyone else have thoughts?

nilsdeppe commented:
I'm with Will on this. Literally the point of metavariables and system is to consolidate the compile time variables in as few places as is reasonable. This PR seems to be going the opposite direction, which we had tried and found very difficult to oversee even at current scales, never mind as more components are added.

nilsvu commented:
Sounds reasonable. I also like the idea of removing the implicit knowledge of the Metavariables from `apply` that Will mentioned. Then the option 3. is what I'd find most clear, but I understand your objections to that and 2. would probably be a good balance between clarity and convenience.

So that means `Metavariables::system` doesn't need to be public anymore, but we'd template actions on `<size_t Dim, ..., typename System>` (if needed), correct? We should also make sure to document precisely what types a `System` is expected to expose.

nilsdeppe commented:
No, use the metavariables. That's literally the point of it. It is the "input file" to the metaprogram. If you start moving things everywhere then horrible inconsistencies will happen between different actions, etc. and people will be presented with compiler errors that have template errors that'll take me 20 minutes to figure out. We did try that route previously and it turned into a real mess, as I've said before.

If something *can* be grabbed from the global cache without knowing the system, that's fine to do because the information for what goes into the global cache comes from the metavariables (mostly, it at least is tied together there). That is what allows for great reduction of possible inconsistent states.

I've still yet to be presented with a reason *why* anyone wants to get rid of the metavariables and make users have to keep different things in sync.

wthrowe commented:
Templating on system might be useful for dual systems.  It also might not, since we have no clue what dual systems are going to look like, so we shouldn't be basing any decisions on that.

My overall take is that we may want some increased templating, but it will be months before we know if we do and if so exactly what we want to change.  I don't think we should go in this direction now unless it actually turns out to be necessary for the elliptic solver.

wthrowe commented:
I had a discussion of this earlier this week with @nilsleiffischer that I found confusing, but I wasn't at the time able to articulate why.  I think I have a better handle on what I was thinking now.

For some background, the goal here is to solve a problem that is traditionally written (schematically) in the form:
* `u^*_0 = u^*(u_0)`
* `v_{1/2} = f(u_0, u^*_0)`
* `v^*_{1/2} = v^*(u_0, v_{1/2})`
* `A u_1 = g(u_0, v_{1/2}, v^*_{1/2})`

Where `A` is a symmetric operator that needs to be inverted.  This requires two calculations of numerical fluxes (`u^*_0` and `v^*_{1/2}`) per step, where one can in principle depend on the other.

@kidder and I previously recommended attempting a treatment of the problem where `v` is treated on an equal footing with `u`.  @nilsleiffischer stated that this produces a non-symmetric operator that has to be inverted, which would be unpleasant.

My confusion arises from the fact that I don't expect this transformation to result in any new operators being introduced.  The form I was envisioning was
* `u^*_0 = u^*(u_0)`
* `v_1 = f(u_0, u^*_0)`
* `v^*_0 = v^*(u_0, v_0)`
* `A u_1 = g(u_0, v_0, v^*_0)`

In these equations, the numerical fluxes do not depend on one another, so they can be communicated at the same time.  (One would evaluate the third equation before the second, now.)  The rewrite only changed the arguments to the functions we already had, so I don't understand how it could have made things harder to evaluate.

Is there some reason this scheme is not feasible?

nilsvu commented:
@wthrowe thanks for looking at this again! The two schemes you outline are different, because in the second you compute `v^*_0` from an (initially even nonexisting) `v_0`, whereas in the first you (correctly) use the previously computed `v_{1/2}`. Perhaps it's possible to make the second scheme work too, but that would mean approximating `v_{1/2}` by `v_0` and I don't know if that's possible. I haven't seen anyone do that, either. I have so far concluded that it's just incorrect.

Right now I plan to do the following (suggestions welcome):
1. **Implement GMRES:** The first-order scheme (where we treat u and v on equal footing, making it a coupled set of first-order PDEs), is not symmetric because it involves first derivatives (it's quite easy to see: a differentiation matrix is not symmetric). Since we'll have to deal with first derivatives anyway (and therefore non-symmetric operators, e.g. for XCTS), I'm currently working on a GMRES linear solver that can deal with that. With that done, no architectural changes should be necessary to solve first-order formulations of elliptic systems (we don't even need the mass operator of #737).
2. **Implement primal formulation:** Eventually we'll surely want to get rid of the auxiliary variables, since they increase the number of degrees of freedom significantly (double them in 1D, worse in higher dimension). This is remedied in the primal formulation, for which we (only) need to implement another flux-lifting scheme. This may also be useful for the evolution code. I think it would be good to keep this in mind when changes to the flux communication code are made, so that we don't couple the communication code too closely to the assumption of always having N.F and N.F^*.

nilsvu commented:
@wthrowe Regarding the fluxcomm code, I think it can be very easily disentangled from the flux lifting (if this becomes necessary at some point): In addition to the `packaged_data` it also stores the N.F into the `local_data` right now, only so that e.g. `ApplyBoundaryFluxesGlobalTimeStepping` can pass the `local_data` to `lift_flux` which needs the N.F. But with my recent PR where we made the `package_data` arguments explicit, we can now just delete that part and only store in `local_data` what's needed for the numerical flux computation. Then in `lift_flux` we can just get what's needed directly out of the databox. That also includes the `MagnitudeOfFaceNormal` btw.

wthrowe commented:
The local time-stepping version of flux processing performs (roughly) an Adams-Bashforth integration, so it needs past data.  So it can't pull anything time-dependent from the DataBox.

nilsvu commented:
We can still just delete line 268 in
https://github.com/sxs-collaboration/spectre/blob/8be65c0f4dd41adc021bef5bd7a0c4f1fdc2528b/src/NumericalAlgorithms/DiscontinuousGalerkin/Actions/FluxCommunication.hpp#L268 since it‘s not needed for the flux communication, and instead move the functionality to the flux processing, or am I missing something?

nilsdeppe commented:
Can we close this? It seems like there is little desire to have this implementation and that an alternative one should be found.

nilsvu commented:
Yes, I kept this open for the ongoing discussion. Will open an issue instead.
